blockchain/chain_sync/src/sync.rs

```diff
@@ -461,13 +461,46 @@ where
         }
         Ok(())
     }
-
-    async fn sync_fork(&mut self, _head: &Tipset, _to: &Tipset) -> Result<Vec<Arc<Tipset>>, Error> {
-        // TODO sync fork until tipsets are equal or reaches genesis
-        todo!()
+    /// fork detected, collect tipsets to be included in return_set sync_headers_reverse
+    async fn sync_fork(&mut self, head: &Tipset, to: &Tipset) -> Result<Vec<Tipset>, Error> {
+        // TODO change from using random peerID to managed
+        let peer_id = PeerId::random();
+        // pulled from Lotus: https://github.com/filecoin-project/lotus/blob/master/chain/sync.go#L996
+        const FORK_LENGTH_THRESHOLD: u64 = 500;
+
+        // Load blocks from network using blocksync
+        let tips: Vec<Tipset> = self
+            .network
+            .blocksync_headers(peer_id.clone(), head.parents(), FORK_LENGTH_THRESHOLD)
+            .await
+            .map_err(|_| Error::Other("Could not retrieve tipset".to_string()))?;
+
+        let mut ts = self.chain_store.tipset_from_keys(to.parents())?;
+
+        for i in 0..tips.len() {
+            if *ts.tip_epoch().chain_epoch() == 0 {
+                if self.genesis.blocks()[0] != ts.blocks()[0] {
```

Comment:
```suggestion
                if self.genesis.blocks() != ts.blocks() {
```
Reason why comparison of only first header?

Reply:
I figured a genesis will only be one header but the above would work fine.

Reply:
idk it's just safer to make sure also that the tipset you're checking also doesn't have any other headers. Extremely unlikely edge case but the logic seems incorrect, maybe just to me?

Reply:
Your above suggestion covers the edge case and its actual purpose. Change made.